myia/hypermap.py

```diff
@@ -367,10 +382,19 @@ def _leafcall(args):
         elif is_dataclass(main):
             results = {}
             for name, field in main.__dataclass_fields__.items():
-                args = [
-                    getattr(x, name) if nonleaf else x for x, nonleaf in argmap
-                ]
-                results[name] = _reccall(args)
+                if (
+                    self.update_tag
+                    and not main.__annotations__[name] is self.update_tag
+                ):
+                    (arg0, nonleaf), *_ = argmap
+                    assert nonleaf
```

Comment:
Hi @breuleux , sorry, I forgot to review this PR ! It's ok for me ! Just a question: why arg0 must not be a leaf ? (same above at line 259)

Reply:
That's because `arg0` is where we get the data for fields that don't have the update tag, and leaves don't have these fields by definition.

So if we only update field `x` of `Point(x, y)`, this means we allow `Point(x, y) + z == Point(x + z, y)` but we don't allow `z + Point(x, y)`. In theory, we *could* allow it, but I think it is better (and simpler) to always get unchanged fields from the first argument than to have a more complicated rule like "it comes from the first non-leaf argument." We can always revise this if there is a need.

Reply:
Ok !